compiler/quilt/vfs.py

```diff
@@ -106,31 +106,73 @@ def open_func(filename, mode='r', module=module, *args, **kwargs):
     patcher.start()
     return patcher
 
+
+# Note: To test this, uninstall h5py, then run your quilt_vfs_test.py.
+def scrub_patchmap(patchmap, verbose=False, strict=False):
+    """Return a version of patchmap with missing modules/callables removed.
+
+    :param verbose: Mention each dropped module and/or function
+    :param strict: Raise an error if module exists but func doesn't
+    """
+    result = {}
+
+    for modname, funcname in patchmap.items():
+        try:
+            module = importlib.import_module(modname)
+        except ImportError:
+            if verbose:
+                print("Dropped {} from the patch map (module not present)".format(modname))
+            continue
+        if not hasattr(module, funcname):
+            if verbose:
+                print("Dropped {}.{} from the patch map (func/class not present)")
+            if strict:
+                raise AttributeError("{} is missing from module {}".format(funcname, modname))
+            continue
+        if not callable(getattr(module, funcname)):
+            if verbose:
+                print("Dropped {}.{} from the patch map (not callable)")
+            if strict:
+                raise TypeError("{}.{} is not callable".format(funcname, modname))
+            continue
+        result[modname] = funcname
+    return result
+
+
 DEFAULT_MODULE_MAPPINGS = {
-    'builtins': 'open', 'bz2': 'BZ2File', 'gzip': 'GzipFile'
-    # for keras/tensorflow
-    ,'h5py': 'File',
+    'builtins': 'open',
+    'bz2': 'BZ2File',
+    'gzip': 'GzipFile',
+    'h5py': 'File',    # for keras/tensorflow
     # manually add tensorflow because it's a heavyweight library
-    #'tensorflow': 'gfile.FastGFile', 'tensorflow': 'gfile.GFile'
+    #'tensorflow': 'gfile.FastGFile', 
+    #'tensorflow': 'gfile.GFile',
```

Comment:
given scrub_patchmap, can you add tensorflow back in?  or does it break on multi-level mappings?

Reply:
It's not multilevel mappings that cause the problem, it's that tensorflow has improper behavior.

For example, if you sucessfully do:
`import tensorflow`
..then tensorflow exists in your current namespace.

If you successfully do `import tensorflow.python`, then tensorflow should exist in your current namespace, and python should exist in the tensorflow namespace.  ..but it doesn't.  tensorflow.python gives jack.

..however, I think I've found a way around that -- but, that's for tomorrow.

Reply:
Killer, can't wait.